traversal/procedure/ProcedureVertex.java

```diff
@@ -37,65 +37,60 @@
 
 import javax.annotation.Nullable;
 import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
 import java.util.TreeSet;
-import java.util.concurrent.atomic.AtomicReference;
 
-import static com.vaticle.typedb.common.collection.Collections.set;
 import static com.vaticle.typedb.common.util.Objects.className;
 import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.ILLEGAL_CAST;
 import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
-import static com.vaticle.typedb.core.common.iterator.sorted.SortedIterators.Forwardable.emptySorted;
-import static com.vaticle.typedb.core.common.iterator.sorted.SortedIterators.Forwardable.iterateSorted;
 import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
 import static com.vaticle.typedb.core.common.iterator.sorted.SortedIterator.ASC;
+import static com.vaticle.typedb.core.common.iterator.sorted.SortedIterators.Forwardable.emptySorted;
+import static com.vaticle.typedb.core.common.iterator.sorted.SortedIterators.Forwardable.iterateSorted;
 import static com.vaticle.typedb.core.graph.common.Encoding.ValueType.STRING;
 import static com.vaticle.typedb.core.graph.common.Encoding.Vertex.Type.ROLE_TYPE;
 import static com.vaticle.typedb.core.traversal.predicate.PredicateOperator.Equality.EQ;
+import static java.util.stream.Collectors.toList;
 
 public abstract class ProcedureVertex<
         VERTEX extends Vertex<?, ?>,
         PROPERTIES extends TraversalVertex.Properties
         > extends TraversalVertex<ProcedureEdge<?, ?>, PROPERTIES> {
 
-    private final boolean isStartingVertex;
-    private final AtomicReference<Set<Integer>> dependedEdgeOrders;
-    private ProcedureEdge<?, ?> branchEdge;
+    private int order;
+    private ProcedureEdge<?, ?> lastInEdge;
+    private List<ProcedureEdge<?, ?>> orderedOuts;
+    private Set<ProcedureEdge<?, ?>> transitiveOuts;
+    private Set<Identifier.Variable> scopesModified;
 
-    ProcedureVertex(Identifier identifier, boolean isStartingVertex) {
+    ProcedureVertex(Identifier identifier) {
         super(identifier);
-        this.isStartingVertex = isStartingVertex;
-        this.dependedEdgeOrders = new AtomicReference<>(null);
     }
 
     public abstract Forwardable<? extends VERTEX, Order.Asc> iterator(GraphManager graphMgr, Traversal.Parameters parameters);
 
     @Override
     public void in(ProcedureEdge<?, ?> edge) {
         super.in(edge);
-        if (branchEdge == null || edge.order() < branchEdge.order()) branchEdge = edge;
+        if (lastInEdge == null || edge.order() > lastInEdge.order()) lastInEdge = edge;
```

Comment:
`lastInEdge` and its consequences can be deleted when the query planner orders vertices natively

cc @dmitrii-ubskii